propolis/src/block.rs

```diff
@@ -53,53 +54,138 @@ pub struct BlockInquiry {
 
         let fp = OpenOptions::new().read(true).write(!is_ro).open(p)?;
-        let is_raw = fp.metadata()?.file_type().is_char_device();
+        let is_ro = readonly || meta.permissions().readonly();
+        let len = fp.metadata().unwrap().len() as usize;
+
+        Ok(FileBlockDevBackingStore { fp: fp, is_ro: is_ro, len: len })
+    }
+
+    pub fn get_file(&self) -> &File {
+        &self.fp
+    }
+}
+
+impl BlockDevBackingStore for FileBlockDevBackingStore {
+    fn issue_read(&self, offset: usize, sz: usize) -> Result<Vec<u8>> {
+        let mut vec = Vec::<u8>::with_capacity(sz);
+
+        let nread = unsafe {
+            libc::pread(
+                self.fp.as_raw_fd(),
+                vec.as_mut_ptr() as *mut libc::c_void,
+                sz,
+                offset as i64,
+            )
+        };
+
+        if nread == -1 {
+            return Err(Error::last_os_error());
+        }
+
+        assert_eq!(nread as usize, sz);
+        Ok(vec)
+    }
+
+    fn issue_write(&self, bytes: &[u8], offset: usize) -> Result<()> {
+        let nwritten = unsafe {
+            libc::pwrite(
+                self.fp.as_raw_fd(),
+                bytes.as_ptr() as *const libc::c_void,
+                bytes.len(),
+                offset as i64,
+            )
+        };
 
-        let mut this = Self {
-            fp,
-            is_ro,
+        if nwritten == -1 {
+            return Err(Error::last_os_error());
+        }
+
+        assert_eq!(nwritten as usize, bytes.len());
+        Ok(())
+    }
+
+    fn issue_flush(&self) -> Result<()> {
+        let res = unsafe { libc::fdatasync(self.fp.as_raw_fd()) };
+
+        if res == -1 {
+            Err(Error::new(ErrorKind::Other, "file flush failed"))
+        } else {
+            Ok(())
+        }
+    }
+
+    fn is_ro(&self) -> bool {
+        self.is_ro
+    }
+
+    fn len(&self) -> usize {
+        self.len
+    }
+}
+
+/// Standard [`BlockDev`] implementation. Requires a backing store.
+pub struct PlainBdev<R: BlockReq, S: BlockDevBackingStore> {
+    backing_store: S,
+    block_size: usize,
+    sectors: usize,
+    reqs: Mutex<VecDeque<R>>,
+    cond: Condvar,
+}
+
+impl<R: BlockReq, S: BlockDevBackingStore> PlainBdev<R, S> {
+    /// Creates a new block device from a device at `path`.
+    pub fn create(backing_store: S) -> Result<Arc<Self>> {
+        let len = backing_store.len();
+
+        let this = Self {
+            backing_store,
             block_size: 512,
-            sectors: 0,
-            is_raw,
+            sectors: len / 512,
             reqs: Mutex::new(VecDeque::new()),
             cond: Condvar::new(),
         };
-        this.raw_init();
 
         Ok(Arc::new(this))
     }
-    fn raw_init(&mut self) {
-        // TODO: query block size, write cache, discard, etc
-        assert!(!self.is_raw);
-        let len = self.fp.metadata().unwrap().len() as usize;
-        self.sectors = len / self.block_size;
-    }
+
+    /*
```

Comment:
Nit: while rust does support block comments, generally line comments are favoured. I don't know if Oxide has any particular style guide but that's what the Rust standard library follows at least: https://github.com/rust-lang/rfcs/blob/master/text/1574-more-api-documentation-conventions.md#use-line-comments

Reply:
Fixed in 285d000.